test/sslapitest.c

```diff
@@ -1865,35 +1871,238 @@ static int ocsp_server_cb(SSL *s, void *arg)
+        return 0;
+
+    SSL_get0_chain_certs(s, &server_certs);
+
+    if (server_certs != NULL && sc != NULL) {
+        /* certificate chain is available */
+        num = sk_X509_num(server_certs);
+    } else if (sc != NULL) {
+        /*
+         * certificate chain is not available,
+         * set num to 1 for server certificate
+         */
+        num = 1;
+    }
+
+    if (*argi == 2)
+        num = 1;
+
+    sk_resp = sk_OCSP_RESPONSE_new_null();
+
+    thisupd = X509_gmtime_adj(NULL, 0);
+    nextupd = X509_time_adj_ex(NULL, 1, 0, NULL);
+
+    for (i = 0; i < num; i++) {
+        bs = OCSP_BASICRESP_new();
+
+        if (i == 0) {
+            /* get OCSP response for server certificate first */
+            ssl_cert = SSL_get_certificate(s);
+        } else {
+            /*
+             * for each certificate in chain (except root) get
+             * the OCSP response
+             */
+            ssl_cert = sk_X509_value(server_certs, i - 1);
+        }
+
+        issuer = sk_X509_value(server_certs, i);
+
+        cert_id = OCSP_cert_to_id(md, ssl_cert, issuer);
+
+        OCSP_basic_add1_status(bs, cert_id, V_OCSP_CERTSTATUS_GOOD,
+                               0, NULL, thisupd, nextupd);
+
+        if (!TEST_ptr(bio_signer_key = BIO_new_file(signer_key_file, "r"))
+            || !TEST_ptr(signer_key = PEM_read_bio_PrivateKey_ex(bio_signer_key, NULL,
+                                                                 NULL, NULL, libctx, NULL)))
+            return SSL_TLSEXT_ERR_ALERT_FATAL;
+
+        if (!TEST_ptr(signer_cert = load_cert_pem(signer_cert_file, libctx))
+            || !TEST_true(OCSP_basic_sign(bs, signer_cert, signer_key, EVP_sha256(),
+                                          NULL, OCSP_NOCERTS)))
+            return SSL_TLSEXT_ERR_ALERT_FATAL;
+
+        if (!TEST_ptr(dummy_ocsp_resp = OCSP_response_create(OCSP_RESPONSE_STATUS_SUCCESSFUL, bs)))
+            return SSL_TLSEXT_ERR_ALERT_FATAL;
+
+        sk_OCSP_RESPONSE_push(sk_resp, dummy_ocsp_resp);
+
+        OCSP_BASICRESP_free(bs);
+        OCSP_CERTID_free(cert_id);
+    }
+
+    num = sk_OCSP_RESPONSE_num(sk_resp);
+
+    (void)SSL_set0_tlsext_status_ocsp_resp_ex(s, sk_resp);
+
+    ocsp_server_called = 1;
+    return SSL_TLSEXT_ERR_OK;
+}
+
+static int ocsp_client_cb_multi(SSL *s, void *arg)
+{
+    int *argi = (int *)arg;
+    STACK_OF(OCSP_RESPONSE) *sk_resp = NULL;
+    OCSP_RESPONSE *resp;
+    STACK_OF(X509) *server_certs = NULL;
+    X509 *ssl_cert = NULL, *issuer_cert = NULL;
+    OCSP_BASICRESP *bs = NULL;
+    OCSP_CERTID *cert_id = NULL;
+    OCSP_SINGLERESP *sr = NULL;
+    OCSP_CERTID *cid = NULL;
+    ASN1_OBJECT *cert_id_md_oid;
+    const EVP_MD *cert_id_md;
+    int i, num;
+    int testresult = 1;
+
+    num = SSL_get0_tlsext_status_ocsp_resp_ex(s, &sk_resp);
+
+    /* check if we get as many OCSP responses as expected */
+    if ((*argi == 1 && num != 3) || (*argi == 2 && num != 1))
+        return 0;
```

Comment:
This appears counter-intuitive. Better use
```suggestion
    if (*argi < 1 || *argi > 2 || num != *argi)
        return 0;
```